rdflib/graph.py

```diff
@@ -1597,6 +1597,70 @@ def add_to_cbd(uri):
 
         return subgraph
 
+    def describe_cbd(self, resource, subgraph):
```

Comment:
See:
- https://github.com/RDFLib/rdflib/issues/1914

Also you will have to include tests specifically for this function for this PR to be merged, but still I don't think it should exist, if `Graph.cbd` does not do what it says in the docstring it is a bug and should be fixed.

Reply:
Hey!

Thanks for your review.

We have included the tests for this describe_cbd function in the test file [test_graph_cbd.py](https://github.com/RDFLib/rdflib/pull/1913/files#diff-a8bea910ef28d48bce0a390d9ab24c55ac32f1c9a39d9cd215224e5447b70c83) as the [testCbdDescribeReified()](https://github.com/RDFLib/rdflib/pull/1913/files#diff-48452d143aa4756150802ff09369f51735226fe016d1f5ad7551b4594f00d57eR114) function. The test example which we have included will be able to cover all the cases for the definition of DESCRIBE query. The test example can be visualized as:-
![Screen Shot 2022-05-12 at 10 14 53 PM](https://user-images.githubusercontent.com/55681714/168126890-0027932f-de20-4d13-936c-b917a2092b19.png)

Reply:
The function **describe_cbd** is a variant of the original CBD implementation. I have modified the function to recursively consider all triples containing the query resource as an **object** and subject. The original function returns the CBD containing triples where the resource is a subject only. However, for implementing the `DESCRIBE` query, we need all triples containing the resource – both as an object and subject. We also ensure that the blank nodes are recursively resolved – this is in line with the definition of `DESCRIBE` query.

Reply:
> We have included the tests for this describe_cbd function in the test file [test_graph_cbd.py](https://github.com/RDFLib/rdflib/pull/1913/files#diff-a8bea910ef28d48bce0a390d9ab24c55ac32f1c9a39d9cd215224e5447b70c83) as the [testCbdDescribeReified()](https://github.com/RDFLib/rdflib/pull/1913/files#diff-48452d143aa4756150802ff09369f51735226fe016d1f5ad7551b4594f00d57eR114) function.

My appologies, in that case you will need tests for SPARQL `DESCRIBE` as part of the PR.

> The function **describe_cbd** is a variant of the original CBD implementation. I have modified the function to recursively consider all triples containing the query resource as an **object** and subject. The original function returns the CBD containing triples where the resource is a subject only.

If it does something different then it should at least have a different docstring which it does not, and then the next question is, why does it do something different. There are one definition of a Concise Bounded Description, and it is this:
https://www.w3.org/Submission/CBD/#definition

> However, for implementing the `DESCRIBE` query, we need all triples containing the resource – both as an object and subject. We also ensure that the blank nodes are recursively resolved – this is in line with the definition of `DESCRIBE` query.

Can you clarify why? As per the SPARQL 1.1 spec:

https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#describe

> The DESCRIBE form returns a single result RDF graph containing RDF data about resources. This data is not prescribed by a SPARQL query, where the query client would need to know the structure of the RDF in the data source, but, instead, is determined by the SPARQL query processor.

And goes on to say:
> Other possible mechanisms for deciding what information to return include Concise Bounded Descriptions [[CBD](https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#CBD)].

It does not state that DESCRIBE must return the CBD, but I do expect that if we return the Concise Bounded Description that we return it as defined, in which case `Graph.cbd` should do it, and if it does not should be fixed to do it.

Reply:
> It does not state that DESCRIBE must return the CBD

`DESCRIBE` should not return *just* the CBD, because as we mentioned below

> However, for implementing the DESCRIBE query, we need all triples containing the resource – both as an object and subject.

This is in line with the definition of the `DESCRIBE` query.

> If it does something different then it should at least have a different docstring which it does not as far as I can tell

Apologies for not providing a separate docstring, I can add that to make it well documented. A few things here and there can be improved wrt documentation, I can check that out. 

Further,

> in that case, you will need tests for SPARQL DESCRIBE as part of the PR.

I have also provided the link to a testing file that has such an example. 

> To show the working of DESCRIBE query, we have provided a [tester file](https://drive.google.com/file/d/1jIbMgmv_hPVbMcJXIRbQmx0-TjcOudax/view?usp=sharing).

If that matches what you expect, we can add that file to the PR. For now, we have just attached the [link](https://drive.google.com/file/d/1jIbMgmv_hPVbMcJXIRbQmx0-TjcOudax/view?usp=sharing).

Reply:
> `DESCRIBE` should not return _just_ the CBD, because as we mentioned below
> 
> > However, for implementing the DESCRIBE query, we need all triples containing the resource – both as an object and subject.

But I don't understand why, where does this requirement come from, could you cite some specification that requires this behavior or indicates why you say it needs "all triples containing the resource – both as an object and subject."? If this is just what you as an individual would prefer I don't think it is sufficient justification.

> I have also provided the link to a testing file that has such an example.

It has to be in our test suite, if it is not then I don't know if it passes (as I won't run it on my computer) and we also can't know if there is a regression.

Reply:
> Apologies for not providing a separate docstring, I can add that to make it well documented. A few things here and there can be improved wrt documentation, I can check that out.

To begin with it will be best to just cite what specification dictates this algorithm/behaviour.

Reply:
fwiw, the result of a "DESCRIBE" query is *determined by the service* and the definition is sufficiently vague as to prompt the W3 to set `description` in double quotes: 

> “The DESCRIBE form takes each of the resources identified in a solution, together with any resources directly named by IRI, and assembles a single RDF graph by taking a "description" which can come from any information available including the target RDF Dataset. The description is determined by the query service.”

The [EBNF for SPARQL `DESCRIBE`](https://www.w3.org/2021/12/rdf-star.html#sparql-star-ebnf) suggests that an implementation that conforms to the spec is going to demand rather more than just defaulting to returning a cbd:

```
DescribeQuery | ::= | "DESCRIBE" ( VarOrIri+  \| "*")  DatasetClause*  WhereClause?  SolutionModifier
```